app/controllers/v3/roles_controller.rb

```diff
@@ -117,15 +113,10 @@ def create_org_role(message)
     unauthorized! unless permission_queryer.can_write_to_active_org?(message.organization_guid)
     suspended! unless permission_queryer.is_org_active?(message.organization_guid)
 
-    if message.user_guid
-      user_guid = message.user_guid
-      unprocessable_user! if user_in_db?(user_guid) && !user_is_readable?(user_guid)
-    else
-      user_guid = lookup_user_guid_in_uaa(message.username, message.user_origin)
-      unprocessable_user! unless user_guid
-    end
+    user_guid = message.user_guid || lookup_user_guid_in_uaa(message.username, message.user_origin)
 
     user = User.first(guid: user_guid) || create_cc_user(user_guid)
+    unprocessable_user! unless user
```

Comment:
Is this even possible, i.e. can `user` be `nil`? Or can this line as well as the `unprocessable_user!` method be removed?

Reply:
good catch. I think to keep the intent with the original code it should read 

```
   unprocessable_user! unless user_guid
   user = User.first(guid: user_guid) || create_cc_user(user_guid)
```

But rereading the code I'm less sure of this change.  Seems like the previous code only created new users if the user already existed in UAA, with this new code it will create new users even if the user doesn't exist at all?

Reply:
I think the code needs to go back to two branches:

```
    if message.user_guid
      user_guid = message.user_guid
      unprocessable_user! if user_in_db?(user_guid)
    else
      user_guid = lookup_user_guid_in_uaa(message.username, message.user_origin)
      unprocessable_user! unless user_guid
    end
  ```

That way when creating a role and supplying a non-existent user_guid it still fails, instead of trying to create a user that doesn't exist in UAA

Reply:
Nah you're right. `User.first(guid: user_guid)` can be nil, but in that case `create_cc_user(user_guid)` won't be. 

I guess that was an attempt to handle the case where `create_cc_user` fails, but probably not necessary since that raises its own errors. I'll cut that now.

Reply:
@will-gant I think throughout the refactor the original purpose of the `unprocessable_user!` call was lost.  I think it's still necessary for cases where the `user_guid` is sent via the message and does not exist in the db

Reply:
> @will-gant I think throughout the refactor the original purpose of the `unprocessable_user!` call was lost. I think it's still necessary for cases where the `user_guid` is sent via the message and does not exist in the db

This case (i.e. provided `user_guid` does not exist in UAA) wasn't handled before.
```
unprocessable_user! if user_in_db?(user_guid) && !user_is_readable?(user_guid)
```
The original check was for the "unaffiliated" user only.

So from my point of view the refactoring is now complete. But the question that remains is, if we might want to add a check (e.g. to `create_cc_user`) if the given `user_guid` is valid (this could leverage [User.uaa_users_info](https://github.com/cloudfoundry/cloud_controller_ng/blob/04aa8e89f3934823245981d0df30079e66bf1a34/app/models/runtime/user.rb#L211-L214)). Currently I'm able to add an arbitrary guid (e.g. `bommel`) and this will be displayed as follows:
```
Getting users in org my-org as me@example.com...

ORG MANAGER
  me@example.com (my-origin)

BILLING MANAGER
  No BILLING MANAGER found

ORG AUDITOR
  bommel (client)
```

Reply:
Ah apologies for the out-of-order reply. As you possibly guessed, I was responding to what Phillip wrote above.

In fact the current behaviour is that we validate that the user exists in UAA only if a _username_ is provided, not a guid:

```ruby
user_guid = lookup_user_guid_in_uaa(message.username, message.user_origin)
unprocessable_user! unless user_guid
```

Where a user _guid_ is provided, we emit `unprocessable_user!` only if the user already exists in the Users table but does not already have an org role in the requesting user's org. There's no check against UAA:

```ruby
unprocessable_user! if user_in_db?(user_guid) && !user_is_readable?(user_guid)
```

That permits the current bizarre behaviour:

```console
$ cf curl -X POST /v3/roles -d '{
>   "type": "organization_user",
>   "relationships": {
>     "user": {
>       "data": {
>         "guid": "not-a-real-guid"
>       }
>     },
>     "organization": {
>       "data": {
>         "guid": "459381c3-355a-4cef-903a-598ebed19757"
>       }
>     }
>   }
> }'
{"guid":"4c67678f-428e-4ac8-8840-80328b244ed1","created_at":"2022-10-05T08:18:41Z","updated_at":"2022-10-05T08:18:41Z","type":"organization_user","relationships":{"user":{"data":{"guid":"not-a-real-guid"}},"organization":{"data":{"guid":"459381c3-355a-4cef-903a-598ebed19757"}},"space":{"data":null}},"links":{"self":{"href":"https://api.cf.dev22.aws.bndl.sapcloud.io/v3/roles/4c67678f-428e-4ac8-8840-80328b244ed1"},"user":{"href":"https://api.cf.dev22.aws.bndl.sapcloud.io/v3/users/not-a-real-guid"},"organization":{"href":"https://api.cf.dev22.aws.bndl.sapcloud.io/v3/organizations/459381c3-355a-4cef-903a-598ebed19757"}}}

$ cf curl /v3/users/not-a-real-guid
{"guid":"not-a-real-guid","created_at":"2022-10-05T08:18:41Z","updated_at":"2022-10-05T08:18:41Z","username":null,"presentation_name":"not-a-real-guid","origin":null,"metadata":{"labels":{},"annotations":{}},"links":{"self":{"href":"https://api.cf.dev22.aws.bndl.sapcloud.io/v3/users/not-a-real-guid"}}}
```

The acceptance criteria in the [original story](https://www.pivotaltracker.com/n/projects/2196383/stories/170061485) and [accompanying flowchart](https://www.pivotaltracker.com/file_attachments/101221754/download?inline=true) make clear this was not intentional, and that a 422 should be returned in this case, so it looks like this was just an oversight.

Reply:
@will-gant So after looking at the original story and the flowchart you referenced, I get the feeling that this is intentional. The acceptance block at the end contains a "Not in UAA - GUID" section showing this "bizarre" behaviour and for @Gerg this was correct. And the fact that the cf cli displays such a "user" as `bommel (client)` (see my comment above) implies that this is some special case we should not change.

Reply:
Small extra thing is I think we don't need to throw `unprocessable!` if `lookup_user_guid_in_uaa` doesn't find a user by their username, as the method [already does that](https://github.com/cloudfoundry/cloud_controller_ng/blob/5054365d7ebe44f31d8ab17f2ef0328a4a08e53e/app/controllers/v3/roles_controller.rb#L195-L197).

Reply:
@philippthun So yeah, from the flowchart it looks like they settled on just checking the CCDB for users when a role was being created by user guid, instead of looking them up in the UAA.

![Screenshot 2022-10-05 at 12 17 12](https://user-images.githubusercontent.com/20523607/194048438-44e2af98-3639-4020-bbe0-632f91b29a2e.png)

It's not totally clear to me, though from [this slack chat](https://cloudfoundry.slack.com/archives/CBNDLQBB5/p1576025705243800) I _think_ they wanted to check the UAA but reckoned the UAA client the CC uses for this didn't have the right authorities.

Reply:
My mistake, I misread that condition and thought that behavior was introduced by the PR, and am surprised to see that it was already present and intentional.  

If that's the case then whether or not we change the behavior is out of scope of the PR , IMO, unless you feel inclined to investigate it further.  Thanks for digging up that history.

Reply:
No worries, and thanks for being so thorough! 

Just a note for posterity, in case someone actually does decide to address this issue in future and reads this PR. I reached out to @Gerg, who worked on this logic in 2019. Although he can't remember why the decision was taken to not check the UAA in the case that a user guid is provided, he speculates that it's due to the fact CF users can represent either a user or client in UAA.

CF users always have a guid, regardless of whether they map to a UAA user or a UAA client. But only CF users that represent a UAA user can have a username (which isn't stored by the CC).

Therefore if a **username** is provided for a CF user, only one call to the UAA, to check for the existence of a UAA user, is necessary.

By contrast the **guid** of a CF user could be either the id of a UAA user or the client_id of a UAA client. You'd need to be prepared to check both of these if you wanted to validate it. From the slack conversation there seemed to be a wariness of the hassle involved in getting the [`cloud_controller_username_lookup`](https://github.com/cloudfoundry/cf-deployment/blob/6ec2aca405f23a2eb32ec4108d3385edcfdb9b22/cf-deployment.yml#L719-L723) the `client.read` authority it'd need for this, and there also seemed to be a concern to minimise UAA queries (though I'd argue most org role assignments will be for UAA users, those extra queries will be infrequent).